File: lua/3.lua
--project euler 3: find the largest prime factor of 600851475143
package.path = "..\\libraries\\?.lua;" .. package.path
require "lib"
local t = os.clock()

local result = lib.max_primefactor(600851475143)

print("result: "..result.." - found in "..(os.clock() - t).." s")

File: cpp/3.cpp
#include <iostream>
#include <math.h>

using namespace std;

template <typename Type>
Type max_primefactor(Type n) {
    if (n <= 2) {
        return n;
    }

    Type test = 2;
    while (n > 1) {
        if (n%test == 0) n /= test;
        else test += 1;
    }

    return test;
}

/* Project Euler 3
 * Find the largest prime factor of 600,851,475,143
 */
int main() {
    long long number = 600851475143;
    int primefactor = static_cast<int>( max_primefactor(number) );

    cout << "Largest prime factor: " << primefactor << "\n";
}
